packages/alfa-rules/src/common/predicate/is-visible.ts

```diff
@@ -6,107 +6,76 @@ import { Refinement } from "@siteimprove/alfa-refinement";
 import { Context } from "@siteimprove/alfa-selector";
 import { Style } from "@siteimprove/alfa-style";
 
+import { isClipped } from "./is-clipped";
+import { isOffscreen } from "./is-offscreen";
 import { isRendered } from "./is-rendered";
+import { isReplaced } from "./is-replaced";
 import { isTransparent } from "./is-transparent";
-import { Cache } from "@siteimprove/alfa-cache";
-const { every } = Iterable;
 
+const { every } = Iterable;
 const { not } = Predicate;
 const { and, or } = Refinement;
-const { isElement, hasName } = Element;
+const { isElement } = Element;
 const { isText } = Text;
 
 export function isVisible(device: Device, context?: Context): Predicate<Node> {
   return and(
     isRendered(device, context),
     not(isTransparent(device, context)),
-    not(and(or(isElement, isText), isClipped(device, context))),
+    not(
+      and(
+        or(isElement, isText),
+        or(isClipped(device, context), isOffscreen(device, context))
+      )
+    ),
     (node) => {
       if (
-        Element.isElement(node) &&
+        isElement(node) &&
         Style.from(node, device, context)
           .computed("visibility")
           .some((visibility) => visibility.value !== "visible")
       ) {
         return false;
       }
 
-      if (Text.isText(node)) {
-        return node.data.trim() !== "";
+      if (isText(node)) {
+        if (node.data.trim() === "") {
+          return false;
+        }
+
+        if (
+          node
+            .parent({
+              flattened: true,
+            })
+            .filter(isElement)
+            .some((element) =>
+              Style.from(element, device, context)
+                .computed("font-size")
+                .some((size) => size.value === 0)
+            )
+        ) {
+          return false;
+        }
       }
 
       return true;
     },
-    // non-replaced elements with no visible child are not visible
+    // Non-replaced elements with no visible children are never visible while
     // replaced elements are assumed to be replaced by something visible.
     not(
       and(
         isElement,
         and(not(isReplaced), (element) =>
           every(
-            element.children({ nested: true, flattened: true }),
+            element.children({
+              nested: true,
+              flattened: true,
+            }),
             not(isVisible(device, context))
```

Comment:
For the sake of readability of `isVisible`, I think we can go further in the subfunctions and extract these as two separate predicates.

Reply:
Agreed, though I'll leave that to a separate task.